File: png_echo/bass12/c/contest.c
/* BASS Simple Console Test, copyright (c) 1999-2001 Ian Luck.
==============================================================
Imports: bass.lib, kernel32.lib, user32.lib, winmm.lib
*/

#include <windows.h>
#include <mmsystem.h>
#include <stdio.h>
#include <conio.h>
#include "bass.h"

/* display error messages */
void Error(char *text) 
{
	printf("Error(%d): %s\n",BASS_ErrorGetCode(),text);
	BASS_Free();
	ExitProcess(0);
}

static DWORD starttime;

/* looping synchronizer, resets the clock */
void CALLBACK LoopSync(HSYNC handle, DWORD channel, DWORD data, DWORD user)
{
	starttime=timeGetTime();
}

void main(int argc, char **argv)
{
	HMUSIC mod;
	HSTREAM str;
	DWORD time,pos,level;
	int a,freq;
	BOOL mono=FALSE;
	int act;

	printf("Simple console mode BASS example : MOD/MPx/OGG/WAV player\n"
			"---------------------------------------------------------\n");

	/* check that BASS 1.2 was loaded */
	if (BASS_GetVersion()!=MAKELONG(1,2)) {
		printf("BASS version 1.2 was not loaded\n");
		return;
	}

	if (argc!=2) {
		printf("\tusage: contest <file>\n");
		return;
	}

	/* setup output - default device, 44100hz, stereo, 16 bits, OGG if available */
	if (!BASS_Init(-1,44100,BASS_DEVICE_OGG,0)
		&& !BASS_Init(-1,44100,0,0))
		Error("Can't initialize device");

	/* try streaming the file */
	if ((str=BASS_StreamCreateFile(FALSE,argv[1],0,0,0))
		|| (str=BASS_StreamCreateURL(argv[1],BASS_STREAM_RESTRATE,0))) {
		/* check if the stream is mono (for the level indicator) */
		mono=BASS_ChannelGetFlags(str)&BASS_SAMPLE_MONO;
		/* set a synchronizer for when the stream reaches the end */
		BASS_ChannelSetSync(str,BASS_SYNC_END,0,&LoopSync,0);
		pos=BASS_StreamGetLength(str);
		if (BASS_StreamGetFilePosition(str,1)!=-1) {
			/* streaming from the internet */
			if (pos)
				printf("streaming internet file [%d bytes]",pos);
			else
				printf("streaming internet file");
		} else
			printf("streaming file [%d bytes]",pos);
	} else {
		/* try loading the MOD (with looping, sensitive ramping, surround sound and calculate the duration) */
		if (!(mod=BASS_MusicLoad(FALSE,argv[1],0,0,BASS_MUSIC_LOOP|BASS_MUSIC_RAMPS|BASS_MUSIC_SURROUND|BASS_MUSIC_CALCLEN)))
			/* not a MOD either */
			Error("Can't play the file");
		/* set a synchronizer for when the MOD reaches the end */
		BASS_ChannelSetSync(mod,BASS_SYNC_END,0,&LoopSync,0);
		printf("playing MOD music \"%s\" [%d orders]",BASS_MusicGetName(mod),BASS_MusicGetLength(mod,FALSE));
		pos=BASS_MusicGetLength(mod,TRUE);
	}

	/* calulate and display the time length */
	if (pos) {
		a=BASS_ChannelGetFlags(str?str:mod);
		BASS_ChannelGetAttributes(str?str:mod,&freq,NULL,NULL);
		pos/=freq*(a&BASS_SAMPLE_MONO?1:2)*(a&BASS_SAMPLE_8BITS?1:2);
		printf(" %d:%02d\n",pos/60,pos%60);
	} else /* no time length available */
		printf("\n");

	BASS_Start();
	if (str)
		BASS_StreamPlay(str,FALSE,BASS_SAMPLE_LOOP);
	else
		BASS_MusicPlay(mod);
	starttime=timeGetTime();

   /* NOTE: some compilers don't support _kbhit */
	while (!_kbhit() && (act=BASS_ChannelIsActive(str?str:mod))) {
		/* display some stuff and wait a bit */
		time=timeGetTime()-starttime;
		level=BASS_ChannelGetLevel(str?str:mod);
		pos=BASS_ChannelGetPosition(str?str:mod);
		if (str)
			printf("pos %09d - time %d:%02d - L ",pos,time/60000,(time/1000)%60);
		else
			printf("pos %03d:%03d - time %d:%02d - L ",LOWORD(pos),HIWORD(pos),time/60000,(time/1000)%60);
		if (act==2) {
			/* internet file stream playback has stalled */
			printf("-- buffering : %05d --",
				BASS_StreamGetFilePosition(str,1)-BASS_StreamGetFilePosition(str,0));
		} else {
			for (a=93;a;a=a*2/3) putchar(LOWORD(level)>=a?'*':'-');
			putchar(' ');
			if (mono)
				for (a=1;a<128;a+=a-(a>>1)) putchar(LOWORD(level)>=a?'*':'-');
			else
				for (a=1;a<128;a+=a-(a>>1)) putchar(HIWORD(level)>=a?'*':'-');
		}
		printf(" R - cpu %.1f%%  \r",BASS_GetCPU());
		Sleep(50);
	}
	printf("                                                                         \r");

	/* get the frequency... and wind it down */
	BASS_ChannelGetAttributes(str?str:mod,&freq,NULL,NULL);
	level=freq/40;
	for (;freq>2000;freq-=level) {
		BASS_ChannelSetAttributes(str?str:mod,freq,-1,-101);
		Sleep(5);
	}

	/* fade-out to avoid a "click" */
	for (a=100;a>=0;a-=5) {
		BASS_SetVolume(a);
		Sleep(1);
	}

	BASS_Free();
}

File: png_echo/bass12/tmt/ConTest/Contest.pas
(*
BASS Simple Console Test, copyright (c) 1999-2001 Ian Luck.
TMT Pascal version by Vadim Bodrov, TMT Development Corp. (support@tmt.com)
*)

uses Strings, Windows, MMSystem, BASS, Keyboard;

var
  starttime: DWORD;

function IntToFixed(val, digits: Longint): string;
var
  s: string;
begin
  s := IntToStr(val);
  while Length(s) < digits do s := '0' + s;
  Result := s;
end;

// display error messages
procedure Error(text: string);
begin
  WriteLn('Error(' + IntToStr(BASS_ErrorGetCode) + '): ' + text);
  BASS_Free;
  Halt(0);
end;

// looping synchronizer, resets the clock
procedure LoopSync(handle: HSYNC; channel, data, user: DWORD); stdcall;
begin
  starttime := timeGetTime;
end;

var
  mus: HMUSIC;
  strm: HSTREAM;
  vol: DWORD;
  time, pos, level: DWORD;
  a, pan: Longint;
  freq: DWORD;
  mono: Boolean;
  tmp: array[0..MAX_PATH] of char;

  function StrMus: DWORD;
  begin
    if strm <> 0 then
      Result := strm
    else
      Result := mus;
  end;

begin
  WriteLn('Simple console mode BASS example : MOD/MP3/WAV player');
  Writeln('-----------------------------------------------------');

  // check that BASS 1.2 was loaded
  if (BASS_GetVersion <> MAKELONG(1, 2)) then
  begin
    Writeln('BASS version 1.2 was not loaded');
    Exit;
  end;
  if (ParamCount <> 1) then
  begin
    WriteLn(#9 + 'usage: contest <file>');
    Exit;
  end;

  // setup output - default device, 44100hz, stereo, 16 bits
  if BASS_Init(-1, 44100, 0, 0) = LONGFALSE then
    Error('Can''t initialize device');

  // try streaming the file
  strm := BASS_StreamCreateFile(LONGFALSE, StrPCopy(tmp, ParamStr(1)), 0, 0, 0);
  if strm = 0 then
    strm := BASS_StreamCreateURL(StrPCopy(tmp, ParamStr(1)), BASS_STREAM_RESTRATE, nil);
  if strm <> 0 then
  begin
    // check if the stream is mono (for the level indicator)
    mono := (BASS_ChannelGetFlags(strm) and BASS_SAMPLE_MONO) > 0;
    // set a synchronizer for when the stream reaches the end
    BASS_ChannelSetSync(strm, BASS_SYNC_END, 0, @LoopSync, 0);
    if BASS_StreamGetFilePosition(strm, 1) <> $FFFFFFFF then
      // streaming from the internet
      Writeln('streaming internet file [', BASS_StreamGetLength(strm), ' bytes]')
    else
      Writeln('streaming file [', BASS_StreamGetLength(strm), ' bytes]');
  end else
  begin
    // try loading the MOD (with looping, sensitive ramping and surround sound)
    mus := BASS_MusicLoad(LONGFALSE, StrPCopy(tmp, ParamStr(1)), 0, 0, BASS_MUSIC_LOOP or BASS_MUSIC_RAMPS or BASS_MUSIC_SURROUND);
    if mus = 0 then
      // not a MOD either
      Error('Can''t play the file');
    // set a synchronizer for when the MOD reaches the end
    BASS_ChannelSetSync(mus, BASS_SYNC_END, 0, @LoopSync, 0);
    Writeln('playing MOD music ', BASS_MusicGetName(mus), ' [', + BASS_MusicGetLength(mus,FALSE), ' orders]');
  end;

  BASS_Start;
  if strm <> 0 then
    BASS_StreamPlay(strm, LONGFALSE, BASS_SAMPLE_LOOP)
  else
    BASS_MusicPlay(mus);
  starttime := timeGetTime;

  while (not KbHit) and (BASS_ChannelIsActive(StrMus) <> 0) do
  begin
    // display some stuff and wait a bit
    time := timeGetTime - starttime;
    level := BASS_ChannelGetLevel(StrMus);
    pos := BASS_ChannelGetPosition(StrMus);

    if (strm <> 0) then
      Write('pos ' + IntToFixed(pos, 9) + ' - time ' + IntToStr(time div 60000) + ':' + IntToFixed((time div 1000) mod 60, 2) + ' - L ')
    else
      Write('pos ' + IntToFixed(LOWORD(pos), 3) + ':' + IntToFixed(HIWORD(pos), 3) + ' - time ' + IntToStr(time div 60000) + ':' + IntToFixed((time div 1000) mod 60, 2) + ' - L ');

    if BASS_ChannelIsActive(StrMus) = 2 then
      // internet file stream playback has stalled
      Write('-- buffering : ', BASS_StreamGetFilePosition(strm, 1) - BASS_StreamGetFilePosition(strm, 0), ' --')
    else
    begin
      a := 93;
      while a > 0 do
      begin
        if LOWORD(level) >= a then
          Write('*')
        else
          Write('-');
        a := a * 2 div 3;
      end;

      Write(' ');
      if mono then
      begin
        a := 1;
        while (a < 128) do
        begin
          if LOWORD(level) >= a then
            Write('*')
          else
            Write('-');
          a +:= a - (a shr 1);
        end;
      end else
      begin
        a := 1;
        while (a < 128) do
        begin
          if HIWORD(level) >= a then
            Write('*')
          else
            Write('-');
          a +:= a - (a shr 1);
        end;
      end;
    end;
    Write(' R - cpu ' + Fix(BASS_GetCPU, 2, 1) + '%  ' + #13);
    Sleep(50);
  end;

  // get the frequency... and wind it down
  BASS_ChannelGetAttributes(StrMus, freq, vol, pan);
  level := freq div 40;
  while (freq > 2000) do
  begin
    BASS_ChannelSetAttributes(StrMus, freq, -1, -101);
    freq -:= level;
    Sleep(5);
  end;

  // fade-out to avoid a "click"
  for a := 20 downto 0 do
  begin
    BASS_SetVolume(a * 5);
    Sleep(1);
  end;

  BASS_Free;
  Sleep(500);
end.
